src/mongo/shell.ts

```diff
@@ -5,120 +5,201 @@
-			window.showErrorMessage(error.toString());
+			// Generally if writing to the process' stdin fails it has already exited
+			// with an error, and we will get notification via its stdout. So delay this long
+			// enough for other notifications to be a given a to fire first.
+			setTimeout(() => this.fireError(error), 500);
 		}
 
-		return await new Promise<string>((c, e) => {
-			let executed = false;
-			// timeout setting specified in seconds. Convert to ms for setTimeout
-			const timeout: number = 1000 * vscode.workspace.getConfiguration().get<number>(ext.settingsKeys.mongoShellTimeout);
-			const handler = setTimeout(
+		return await new Promise<string>((resolve, reject) => {
+			// Start timeout timer
+			let timeout: number = 1000 * vscode.workspace.getConfiguration().get<number>(ext.settingsKeys.mongoShellTimeout);
+			if (timeout <= 0) {
+				// No timeout (Number.MAX_SAFE_INTEGER apparently is not valid, so use a full day)
+				timeout = 24 * 60 * 60 * 1000;
+			}
+			const timeoutHandler = setTimeout(
 				() => {
-					if (!executed) {
-						e(`Timed out executing MongoDB command "${script}"`);
-					}
+					reject(`Timed out executing MongoDB command "${script}". To use a longer timeout, modify the VS Code 'mongo.shell.timeout' setting.`);
 				},
 				timeout);
-			const disposable = this.onResult.event(result => {
-				disposable.dispose();
 
-				if (result && result.code) {
-					if (result.code === 'ENOENT') {
-						result.message = `This functionality requires the Mongo DB shell, but we could not find it. Please make sure it is on your path or you have set the '${ext.settingsKeys.mongoShellPath}' VS Code setting to point to the Mongo shell executable file (not folder). Attempted command: "${this.execPath}"`;
-					}
-
-					e(result);
-				} else {
-					let lines = (<string>result.result).split(os.EOL).filter(line => !!line && line !== 'Type "it" for more');
-					lines = lines[lines.length - 1] === 'Type "it" for more' ? lines.splice(lines.length - 1, 1) : lines;
-					executed = true;
-					c(lines.join(os.EOL));
-				}
+			// Handle result or error from the console (via fireOnResultOrError)
+			const disposable = this._onResult.event((result: CommandResult) => {
+				clearTimeout(timeoutHandler);
 
-				if (handler) {
-					clearTimeout(handler);
-				}
+				// Only the first result or error will be processed, all others ignored
+				disposable.dispose();
+
+				// Give STDOUT/STDERR a chance to empty
+				setTimeout(
+					() => {
+						if (result.result !== undefined && !this._stderrData) {
+							if (this._mongoShell.killed || this._exited) {
+								throw new Error(prematureExitMessage);
+							}
+
+							let lines = result.result.split(os.EOL).filter(line => !!line);
+							if (lines[lines.length - 1] === mongoShowMoreMessage) {
+								// CONSIDER: Ideally we would ask or allow the user to ask for more data
+								lines = lines.splice(0, lines.length - 1);
+								lines.push("(More)");
```

Comment:
This seems like a pretty big behavior change. The shell used to keep typing "it" for the user, correct? Perhaps we should discuss this or file an issue to re-create that behavior before releasing as-is?